feat(ivy): ngtsc template sourcemaps

petebacondarwin opened the pull request:
There is a bug in TypeScript https://github.com/Microsoft/TypeScript/issues/29300 that prevents the `templateUrl` URL from working correctly. It is fixed in TS 3.3. We have a check to fail gracefully if a
version of TS is running that does not include the fix.

For inline templates, only string literals are supported.

The actual mapping from the generated ivy code to the templates needs some work to improve its accuracy and resolution.

mary-poppins commented:
You can preview d2a34f1 at https://pr28055-d2a34f1.ngbuilds.io/.

mary-poppins commented:
You can preview 8ed8779 at https://pr28055-8ed8779.ngbuilds.io/.

mary-poppins commented:
You can preview 30c5bd5 at https://pr28055-30c5bd5.ngbuilds.io/.

mary-poppins commented:
You can preview 329e0eb at https://pr28055-329e0eb.ngbuilds.io/.

mary-poppins commented:
You can preview 4e4d6de at https://pr28055-4e4d6de.ngbuilds.io/.

petebacondarwin commented on packages/compiler/src/ml_parser/lexer.ts:
```diff
@@ -232,8 +238,22 @@ class _Tokenizer {
-    if (this._index >= this._length) {
+  private _advanceToStart(start: number) {
+    try {
+      while (this._index < start) {
```
We have to compute the line and column of the start position. We could do something like split on newlines and then add up the length of each line until we got to the start position, with a bit of fiddling around at the end to ensure we got the correct final value, but it seemed safer, and not much slower to use the same advance mechanism to do this.

petebacondarwin replied:
Good idea. Let me look into that.

petebacondarwin commented on packages/compiler/src/ml_parser/lexer.ts:
```diff
@@ -63,6 +63,30 @@ export interface TokenizeOptions {
    * */
   range?: [number, number];
+  /**
+   * If this text is stored in a JavaScript string, then we have to deal with escape sequences.
```
The addition to the lexer covers all these things, I believe, such as line-continuations, hex chars, unicode (fixed and variable width) chars, and even octals.

Is there something I missed?

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/can_sourcemap_external_templates.ts:
```diff
@@ -0,0 +1,67 @@
+ * mappings that correspond ot the correct line/col pairs for this transformed node.
+ *
+ */
+export function canSourceMapExternalTemplates() {
```
Sure, I'll move and rename.

The function is already memoizing the computed value (in `_canSourceMapExternalTemplates`). I don't think we can make that faster, right?

petebacondarwin commented on packages/compiler-cli/src/ngtsc/util/src/ts_source_map_bug_29300.ts:
```diff
@@ -20,10 +20,9 @@ let _canSourceMapExternalTemplates: boolean|undefined;
  */
-export function canSourceMapExternalTemplates() {
-  if (_canSourceMapExternalTemplates === undefined) {
+export function tsSourceMapBug29300Present() {
```
Doh! I renamed it incorrectly!

alxhub commented on packages/compiler/src/output/output_jit.ts:
```diff
@@ -13,37 +13,45 @@ import {EmitterVisitorContext} from './abstract_emitter';
+/**
+ * A helper class to manage the evaluation of JIT generated code.
+ */
+export class JitEvaluator {
```
I would prefer we not make this change, actually. I still weighing the benefits but right now I'm inclined to avoid the use of spying in tests, in favor of proper interface/implementation decoupling.

I've found having tests based on spies negates a large part of the benefits of static typing. In particular, it's incredibly powerful to be able to make an API change in some part of the codebase and see, via TS errors, all of the sites where the API is consumed that need to be updated. Having spies in tests (which are not strongly typed to the API they're spying on) defeats this approach, and leads to unexpected and sometimes hard to debug failures.

I encountered this recently when changing some of the ngtsc APIs that were consumed by ngcc. I was able to fix usages inside ngcc fairly easily, but the tests took me much longer because I wasn't able to rely on TypeScript to resolve any API mismatches.

I would favor a `Jitter` service we could create a proper fake implementation for instead of organizing the functions to be easily spied on.

petebacondarwin replied:
Let me look into that.

alxhub commented on packages/core/src/render3/jit/directive.ts:
```diff
@@ -71,7 +71,8 @@ export function compileComponent(type: Type<any>, metadata: Component): void {
           viewProviders: metadata.viewProviders || null,
         };
-        ngComponentDef = compiler.compileComponent(angularCoreEnv, sourceMapUrl, meta);
+        const templateUrl = metadata.templateUrl || `ng:///${renderStringify(type)}/template.html`;
```
Can we get rid of `sourceMapUrl` above?

petebacondarwin replied:
Oh yeah!

alxhub commented on packages/compiler/src/jit_compiler_facade.ts:
```diff
@@ -140,8 +140,7 @@ export class CompilerFacadeImpl implements CompilerFacade {
     const preStatements = [...constantPool.statements, ...res.statements];
-
-    return jitExpression(res.expression, angularCoreEnv, sourceMapUrl, preStatements);
+    return jitExpression(res.expression, angularCoreEnv, `ng:///${facade.name}.js`, preStatements);
```
Could this commit be merged into the previous? Also, same comment about `sourceMapUrl`.

petebacondarwin replied:
I can squash the commit but these two commits are doing different things.

In this case the `sourceMapUrl` is different to the line being changed. It is the URL of the source template, whereas this line is defining the URL of the generated factory.

alxhub commented:
Wow. Just wow. This is incredible work, sir.

I only have one other remark for you. To keep the parts of the VE compiler that you're changing in sync between master & patch, we should split this PR in two. The refactoring commits that are applicable to the patch branch we should submit in a `master & patch` PR, and the rest of this can land on top of it in `master` only.

👍

petebacondarwin commented:
@alxhub I have refactored the lexer so that you pass in the line and col when providing a range, so that avoids having to advance over the whole file! See ea51273600778af44d7e232955c1f4aa7610165e and 09b13f7def03935195291659a8f2a37ae364e347

petebacondarwin commented:
@alxhub - regarding backporting the VE stuff to patch... let's get this PR into shape and land it. Then I will create a new PR for patch with the best bits cherry-picked.

petebacondarwin commented:
Fixed up, rebased and ready for @alxhub to PTAL. I moved all the fixup commits to directly after the commit they are fixing, to ensure that they will squash correctly when merged into master. So when reviewing you might need to look down the list of commits to find the fixups.

petebacondarwin commented:
Hmm, except the AOT tests are failing...

petebacondarwin committed:
refactor(compiler): use `options` argument for parsers

This commit consolidates the options that can modify the
parsing of text (e.g. HTML, Angular templates, CSS, i18n)
into an AST for further processing into a single `options`
hash.

This makes the code cleaner and more readable, but also
enables us to support further options to parsing without
triggering wide ranging changes to code that should not
be affected by these new options.  Specifically, it will let
us pass information about the placement of a template
that is being parsed in its containing file, which is essential
for accurate SourceMap processing.

petebacondarwin committed:
docs(core): tidy up the description of `resolveComponentResources()`

There were a number of typos and some of the sentences did not
read well.

petebacondarwin committed:
feat(compiler): support tokenizing a sub-section of an input string

The lexer that does the tokenizing can now process only a part the source
string, by passing a `range` property in the `options` argument. The
locations of the nodes that are tokenized will now take into account the
position of the span in the context of the original source string.

This `range` option is, in turn, exposed from the template parser as well.

Being able to process parts of files helps to enable SourceMap support
when compiling inline component templates.

petebacondarwin committed:
refactor(compiler): wrap the jit evaluation in an injectable class

When testing JIT code, it is useful to be able to access the
generated JIT source. Previously this is done by spying on the
global `Function` object, to capture the code when it is being
evaluated. This is problematic because you can only capture
the body of the function, and not the arguments, which messes
up line and column positions for source mapping for instance.

Now the code that generates and then evaluates JIT code is
wrapped in a `JitEvaluator` class, making it possible to provide
a mock implementation that can capture the generated source of
the function passed to `executeFunction(fn: Function, args: any[])`.

petebacondarwin committed:
build: show no warning for large git repos

This warning pops up every time you try to run a node debug
session via bazel. It is not important.

petebacondarwin committed:
feat(compiler): support tokenizing escaped strings

In order to support source mapping of templates, we need
to be able to tokenize the template in its original context.
When the template is defined inline as a JavaScript string
in a TS/JS source file, the tokenizer must be able to handle
string escape sequences, such as `\n` and `\"` as they
appear in the original source file.

This commit teaches the lexer how to unescape these
sequences, but only when the `escapedString` option is
set to true.  Otherwise there is no change to the tokenizing
behaviour.

petebacondarwin committed:
refactor(core): do not remove `templateUrl` when resolving

When we resolve a component `templateUrl` we copy the contents of the
resolved template file into the `template` property.

Previously we would then remove the `templateUrl` to indicate that
the component has been resolved. But this meant that we no longer had
access to the URL of the original template file. This is essential for
diagnostics messages about the template compilation.

Now the existence of the `template` property overrides the existence of
`templateUrl`, which allows us to keep the `templateUrl` property.

petebacondarwin committed:
refactor(compiler): remove unnecessary `!` operators from lexer

When we added the strict null checks, the lexer had some `!`
operators added to prevent the compilation from failing.

This commit resolves this problem correctly and removes the
hacks.

Also the comment

```
// Note: this is always lowercase!
```

has been removed as it is no longer true.

See #24571

petebacondarwin committed:
fix(compiler): support `sourceMappingURL` comments that have trailing lines

Previously the call to `extractSourceMap()` would only work if the
`//#sourceMappingURL ...` was the last line of the file. This doesn't
work if the code is JIT evaluated as the comment is actually the last
line in the body of a function, wrapped by curly-braces.

petebacondarwin committed:
test(core): update JIT source mapping tests for ivy

There are some differences in how ivy maps template source
compared to View Engine.  In this commit we recreate the View Engine
tests for ivy.

petebacondarwin committed:
refactor(compiler): capture `sourceSpan` when converting action bindings to output AST

The `convertActionBinding()` now accepts an optional `baseSourceSpan`,
which is the start point of the action expression being converted in the
original source code.  This is used to compute the original position of
the output AST nodes.

petebacondarwin committed:
fix(core): use the correct generated URL for JIT compiled components

Previously the generated code was being mapped to the `templateUrl`
value.

petebacondarwin committed:
fix(compiler): ensure that event handlers have the correct source spans

When template bindings are being parsed the event handlers
were receiving a source span that included the whole attribute.

Now they get a span that is focussed on the handler itself.

petebacondarwin committed:
fix(compiler): markup lexer should not capture quotes in attribute value

When tokenizing markup (e.g. HTML) element attributes
can have quoted or unquoted values (e.g. `a=b` or `a="b"`).
The `ATTR_VALUE` tokens were capturing the quotes, which
was inconsistent and also affected source-mapping.

Now the tokenizer captures additional `ATTR_QUOTE` tokens,
which the HTML related parsers understand and factor into their
token parsing.

petebacondarwin committed:
fix(core): use the correct template URL in render3 JIT compilation

Previously JIT compiled components did not use the correct URL if
the template was resolved from a `templateUrl`.

petebacondarwin committed:
feat(ivy): add source mappings to compiled Angular templates

During analysis, the `ComponentDecoratorHandler` passes the component
template to the `parseTemplate()` function. Previously, there was little or
no information about the original source file, where the template is found,
passed when calling this function.

Now, we correctly compute the URL of the source of the template, both
for external `templateUrl` and in-line `template` cases. Further in the
in-line template case we compute the character range of the template
in its containing source file; *but only in the case that the template is
a simple string literal*. If the template is actually a dynamic value like
an interpolated string or a function call, then we do not try to add the
originating source file information.

The translator that converts Ivy AST nodes to TypeScript now adds these
template specific source mappings, which account for the file where
the template was found, to the templates to support stepping through the
template creation and update code when debugging an Angular application.

Note that some versions of TypeScript have a bug which means they cannot
support external template source-maps. We check for this via the
`canSourceMapExternalTemplates()` helper function and avoid trying to
add template mappings to external templates if not supported.

petebacondarwin committed:
test(ivy): add template source mapping tests

gkalpak commented on packages/core/test/linker/source_map_integration_node_only_spec.ts:
```diff
@@ -7,92 +7,218 @@
-      // TODO: Figure out a better way to retrieve the JIT sources, without spying on `Function`.
-      const originalProto = ɵglobal.Function.prototype;
-      jitSpy = spyOn(ɵglobal, 'Function').and.callThrough();
-      ɵglobal.Function.prototype = originalProto;
```
🎉

gkalpak reviewed (approved):
This is so great 👌 ❤️ 🎊

petebacondarwin commented on .vscode/settings.json:
```diff
@@ -15,4 +15,5 @@
     "**/bazel-out": true,
     "**/dist": true,
   },
+  "git.ignoreLimitWarning": true,
```
When debugging a Bazel node test, VS code notices that the whole Bazel test output code, through whose source you are debugging has a git repository of its own - but it seems to think that there are thousands of changes...

petebacondarwin replied:
![screenshot 2019-02-11 at 19 58 17](https://user-images.githubusercontent.com/15655/52589748-799e3080-2e37-11e9-88d2-4c298bed9777.png)

petebacondarwin commented on packages/compiler/src/render3/r3_jit.ts:
```diff
@@ -69,6 +69,6 @@ export function jitExpression(
   ];
 
-  const res = jitStatements(sourceUrl, statements, new R3JitReflector(context), false);
+  const res = JitEvaluator.jitStatements(sourceUrl, statements, new R3JitReflector(context), true);
```
Yes, this boolean flag turns on SourceMap generation. :-)

alxhub replied:
Can you add a comment with the parameter name of the flag? `, /* enableSourceMaps */ true`

petebacondarwin replied:
Done - but not in this file (because this code got moved) - see https://github.com/angular/angular/pull/28055/files#diff-a6e5b540987d60524fba9e53e41f2920R175

kara commented:
@petebacondarwin Looks like this just needs a rebase

kara commented:
@petebacondarwin FYI I rebased the PR. 

[presubmit](http://test/OCL:233648522:BASE:233648601:1550000653135:795fa5b8)

mhevery commented:
@petebacondarwin This PR can't be merged because it does not squash cleanly. Could you squash and force push, and I will merge it.
